---
title: "Introduction into machine learning and analyzes of Breast Cancer Proteomes"
author:
  - name: Mats Slik
    affiliation: Hanze University of Applied Sciences
    address:
    - Zernikeplein 11
    - room D1.12
    - email: m.p.slik@st.hanze.nl

date: September 15th, 2022
output:
    pdf_document:
      toc: true
      toc_depth: 2
---


# Dataset



## About Dataset

### information about the data set and the three given files :
**Context:**
This data set contains published iTRAQ proteome profiling of 77 breast cancer samples generated by the Clinical Proteomic Tumor Analysis Consortium (NCI/NIH). It contains expression values for ~12.000 proteins for each sample, with missing values present when a given protein could not be quantified in a given sample.
this data was sampled from 105 originaly from the TCGA (The Cancer Genome Atlas Program - NCI), which was further filterd to a 77 samples containing high quality protein expression samples.
**Content:**

* **File:** 77cancerproteomesCPTACitraq.csv

  + **RefSeqaccessionnumber:** RefSeq protein ID (each protein has a unique
  ID in a RefSeq database)
  + **gene_symbol:** a symbol unique to each gene (every protein is encoded
  by some gene)
  + **gene_name:** a full name of that gene
  + **Remaining columns:** log2 iTRAQ ratios for each sample (protein
  expression data, most important), three last columns are from healthy
  individuals


* **File:** clinicaldatabreast_cancer.csv

  + **First column** "Complete TCGA ID" is used to match the sample IDs in the main cancer proteomes file (see example script).
  + **All other columns** have self-explanatory names, contain data about the cancer classification of a given sample using different methods. 'PAM50 mRNA' classification is being used in the example script.

* **File:** PAM50_proteins.csv

    * **Contains**  the list of genes and proteins used by the PAM50 classification system. The column RefSeqProteinID contains the protein IDs that can be matched with the IDs in the main protein expression data set.

**Past Research:**
Original research paper: https://www.researchgate.net/publication/303509927_Proteogenomics_connects_somatic_mutations_to_signaling_in_breast_cancer

**Summary:** the data were used to assess how the mutations in the DNA are affecting the protein expression landscape in breast cancer. Genes in our DNA are first transcribed into RNA molecules which then are translated into proteins. Changing the information content of DNA has impact on the behavior of the proteome, which is the main functional unit of cells, taking care of cell division, DNA repair, enzymatic reactions and signaling etc. They performed K-means clustering on the protein data to divide the breast cancer patients into sub-types, each having unique protein expression signature. They found that the best clustering was achieved using 3 clusters (original PAM50 gene set yields four different subtypes using RNA data).
Are there different ways to categorize breast cancer based on protein expression data, with machine learning being used to classify them as benign or malignant?

```{r setup, include = 0}
# packages
library(pander)
library(tidyr)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(stringr)
library(hrbrthemes)
library(viridis)

```
## Exploratory Data Analysis

loading of the dataframes and showing the successful loading and its dimensions.
note only the first 5 columns of "77_cancer_proteomes_CPTAC_itraq.csv" are shown since after column 4 they are the same type.
```{r}
proteomes_data <- read.csv(file = "Data//77_cancer_proteomes_CPTAC_itraq.csv")

clinical_data <- read.csv(file = "Data/clinical_data_breast_cancer.csv")
pam50_protein_data <- read.csv(file = "Data/PAM50_proteins.csv")

# showing succeseful loading of data

# only showing first 5 columns of proteomes
pander(head(proteomes_data[1:5], n = 5))
pander(head(clinical_data, n=5))
pander(head(pam50_protein_data, n=5))

# showing the structure/dimensions of dataframe
cat("77_cancer_proteomes_CPTAC_itraq [ number of rows:", nrow(proteomes_data), "number of columns:",ncol(proteomes_data),'\n')
cat("clinical_data [ number of rows:", nrow(clinical_data), "number of columns:",ncol(clinical_data),'\n')
cat("pam50_protein_data [ number of rows:", nrow(pam50_protein_data), "number of columns:",ncol(pam50_protein_data),'\n')

```
Checking if the Proteomes data has been correctly read.
```{r}
str(proteomes_data)
```
Nothing strange about the Proteomes dat everything seems to be read correct.

Checking if the clinical data  has been correctly read.
```{r}
str(clinical_data)
```
Nothing strange about the clinical data everything seems to be read correct.

Checking if the pam50 protein data  has been correctly read.
```{r}
str(pam50_protein_data)
```
Nothing strange about the pam50 protein data everything seems to be read correct.


### codebook

loading of the created codebooks for the three dataframes.
showing also its contents and successful loading
```{r}

cancer_proteomes_CPTAC_codebook <- read.csv2("Data/77_cancer_proteomes_CPTAC_codebook.txt")
clinical_data_codebook <- read.csv2("Data/clinical_data_breast_cancer_codebook.txt")
PAM50_protein_codebook <- read.csv2("Data/PAM50_protein_codebook.txt", sep = ";")


pander(cancer_proteomes_CPTAC_codebook)
pander(clinical_data_codebook)
pander(PAM50_protein_codebook)
```

## Data observation

there are 12553 rows in the data, these are proteins identifiable with a RefSeq ID number and have 86 columns of witch the last 83 are samples(with named with their identifiers and the last three from healthy individuals.
to further use the data i shall reshape it to make the rows samples and each column a protein

## Data cleaning and altering

### altering sample names

the alteration of sample names to corespondent to the clinical data names is needed for further comparison and analyses.
this is done by changing the column names to that of the same format of the clinical data.
this is done with some regex magic
```{r}
# storing a list of the column names
column_names <- names(proteomes_data)

# function
change_sample_name <- function (x){
  #search for TCGA name,if found split and make new name
    if(grepl("TCGA",x) == TRUE){
      temp_list <- as.list(strsplit(x, '[_|-|.]')[[1]])
      x <- str_c(c('TCGA',temp_list[[1]],temp_list[[2]]),collapse = '-')
    }
return (x)
}

# changing of the colnames
colnames(proteomes_data) <- lapply(column_names, change_sample_name)
cat("Old name:",column_names[[4]],",New name:",names(proteomes_data)[[4]])

```
this output show to conversion has been succesfull

### numerical data frame
now  we need to make a data frame with only the numerical data for the ease of analyzes
```{r}
# first making a data frame with only the numerical data, samples start at column number 4 til the end column number 86
proteomes_data_numerical <- proteomes_data[4:86]
```


### Transposing
transposing the created data frame "proteomes_data_numerical", and adding the refseq ID as column name
```{r}
proteomes_data_numerical_transposed <- as.data.frame(t(proteomes_data_numerical))
colnames(proteomes_data_numerical_transposed) <- proteomes_data$RefSeq_accession_number

# checking is succesfull
cat("proteomes_data_numerical_transposed number of rows:", nrow(proteomes_data_numerical_transposed),
    "number of columns:",ncol(proteomes_data_numerical_transposed),'\n')


# lets add a healthy group and cancer, the last 3 columns where from healtyh poeple
# *see content section for the reference
proteomes_data_numerical_transposed$sample_type <- "Cancer"
proteomes_data_numerical_transposed$sample_type[81:83] <- "Healthy"

```
### cleaning

since there are NA values in the data lets see how much
```{r}
count_na_func <- function(x) sum(is.na(x))
# getting NA values per RefSeqID(column)
Na_per_col <- sapply(proteomes_data_numerical_transposed,count_na_func)

hist(Na_per_col, breaks = 40,
     xlab = "Number of NA",
     ylab = "Frequency of columns",
     main = "Frequency of number of NA values per RefSeqID",
     col = "green")

hist(which(Na_per_col>0), breaks = 40,
     xlab = "Number of NA",
     ylab = "Frequency of columns",
     main = "Frequency of number of NA values per RefSeqID with 0 omitted",
     col = "green")
```
```{r}
cat("number of proteins with NA values in them:", sum(Na_per_col > 0), '\n' )

# TODO  decide how many NA values are permitted per protein
proteomes_filtered_data <- proteomes_data_numerical_transposed[Na_per_col < 8]
cat("number of proteins with 8 or more NA values in them and deleted from data:",
    sum(Na_per_col > 8), '\n')

cat("proteomes_filtered_data[number of rows:",
    nrow(proteomes_filtered_data),
    "number of columns:",
    ncol(proteomes_filtered_data),'\n')
```
as we can see from te reports generated by the code we can see that the filtereing of NA was succesfull.

### Merging clinical and protein expression dataframes

to be able to use the Clinical dat we need to merge it to its corresponding row and sample in the Protein expressions
```{r}
# firs assigning row names to clinical data
rownames(clinical_data) <- clinical_data$Complete.TCGA.ID

# removing the used ID column to simplifie it since it has been become redundend
clinical_data <- clinical_data[,-1]

# mergin the two data frames according to the row names (TCGA Identification number),
merged_data <- merge(select(clinical_data, Tumor, Tumor..T1.Coded, AJCC.Stage, Vital.Status), proteomes_data_numerical_transposed, by = 0)

cat("merged_data of rows:", nrow(merged_data),
    "number of columns:",ncol(merged_data),'\n')

```
we can see that not every sample had an entry in the clinical data, so we end up with 6 rows of sample data that get left out of the merged data set


Showing some examples of distributions of protein expression data since there are around 12 000 proteins found
```{r}
boxplot(merged_data[6:76],
        col = rainbow(ncol(merged_data[6:76])),
                           xlab = "Protein",
                           ylab = "log2 iTRAQ ratio",
                           main = "distribution of Protein expression for first 70 Proteins",
                           show.names= FALSE)
```
Since i cant be sure of the significante of every protein i cant simple trow away any sample

to fix this i shall get the standard deviation from every protein

```{r}
fg <- colSds(as.matrix(merged_data[sapply(merged_data, is.numeric)]))
```
```{r}
# assigning as factors
merged_data$Tumor <- factor(merged_data$Tumor)
merged_data$Vital.Status <- factor(merged_data$Vital.Status)
merged_data$AJCC.Stage <- factor(merged_data$AJCC.Stage)

```
```{r}

```
```{r}
write.csv(merged_data, "Data\\export_data.csv", row.names = TRUE)
```
